design/service-apis-implementation.md

```diff
@@ -0,0 +1,258 @@
+
+#### GatewayClass
+
+Contour does not watch or interact with GatewayClass in any way.
+
+#### Gateway
+Contour will watch its configured Gateway object.
+
+Contour is not able to specify the address that the Envoy deployment will listen on, and so will ignore any entries in the Gateway's `spec.addresses` field.
+Contour will not update the `status.addresses` field.
+
+For Contour, the key Gateway section are the Listeners. These define how an implementation should listen for traffic described by Routes.
+
+(The remainder of this document assumes that [#3263](https://github.com/projectcontour/contour/pull/3263) has been approved and merged.)
+
+For Listeners:
+- Listeners inside Gateways will be merged where possible.
+- Listeners that specify a port that the Contour controller considers the default secure port will have a HTTP->HTTPS redirect created for them.
+- Listeners that specify a port that is not exposed via the Envoy service will be rejected.
+- Conflicts within a Gateway will result in the relevant Listeners both being rejected (as there is no way to determine which one was first).
+- Listeners are considered mergable if all the fields out of `hostname`, `port`, and `protocol` match, with some additional rules around TLS.
+- Further merging rules are specified in the detailed design below.
+- Listeners that refer to any other Route than HTTPRoute or TLSRoute will be ignored, and a condition placed on the corresponding `status.listeners[]` object saying that it was ignored because those objects are not supported.
+
+The output of this watcher is both Envoy configuration, and a list of kind/name/namespace details for Routes to watch.
+
+Contour will only update the `status` of Gateway objects.
+
+### HTTPRoute
+Contour will watch all HTTPRoute objects, and filter for entries matching a label selector in its configured Gateway's `spec.listeners[].routes`,
+and will configure the associated routes.
+Configuration of HTTPRoutes will be subject to the rules around the `RouteGateways` field.
+
+TODO: Details about RouteTLSConfig.
+
+Contour will only ever update the `status` of HTTPRoute objects.
+
+Errors or conflicts here will render that section of the config invalid.
+Other valid sections will still be passed to Envoy.
+For each invalid section, Contour will update status information with the section and the reason.
+
+The output of this watcher is Envoy configuration.
+
+### TLSRoute
+Contour will watch all TLSRoute objects, and filter for entries matching a label selector in its configured Gateway's `spec.listeners[].routes`,
+and will configure the associated routes.
+Configuration of TLSRoutes  will be subject to the rules around the `RouteGateways` field.
+
+Contour will only ever update the `status` of TLSRoute objects.
+
+Errors or conflicts here will render that section of the config invalid.
+Other valid sections will still be passed to Envoy.
+
+The output of this watcher is Envoy configuration.
+
+## Detailed Design
```

Comment:
do we want to have one DAG processor for all service-apis objects? One per object type (e.g. `GatewayProcessor`, `HTTPRouteProcessor`, etc)? Something else?

Reply:
I'm not sure. I think that splitting the processors up would be a bit different to how we've done things before, but that's not necessarily bad. It's probably worth the implementor giving this a go.

Reply:
:+1: